docs/design.md

```diff
@@ -0,0 +1,258 @@
+title: sops-compliance-checker configuration
+type: object
+```
+
+This is highly influenced by the [JSON Schema Specification][jsonschema-spec]
+itself to allow for great flexibility in the rule composition.
+
+### Configuration Example
+
+Given the following requirements:
+
+- The SOPS file must contain all necessary approved trust anchors:
+  - AGE recovery key
+    `age1u79ltfzz5k79ex4mpl3r76p2532xex4mpl3z7vttctudr6gedn6ex4mpl3` used for
+    disaster recovery.
+  - AWS KMS keys of **at least one team** owning the component. Additional team
+    keys are allowed, as long as they are part of the list of approved trust
+    anchors.
+  - AWS KMS deploy keys for **exactly one** target environment
+    (development/staging/production).
+- AWS KMS keys **must come in pairs** of two different regions to increase availability.
+- Any excess trust anchors not matching any of these rules **must be rejected**.
+
+One possible configuration to acheive this could be:
+
+```yaml
+---
+rules:
+  # All rules must match. This will automatically reject excess trust anchors
+  # not matching any of the nested rules.
+  - allOf:
+      # Disaster recovery key must be present.
+      - match: age1u79ltfzz5k79ex4mpl3r76p2532xex4mpl3z7vttctudr6gedn6ex4mpl3
+      # The AWS KMS key pair of at least one team must be present.
+      - anyOf:
+          # Regional keys of team-foo.
+          - allOf:
+              - match: arn:aws:kms:eu-central-1:123456789012:alias/team-foo
+              - match: arn:aws:kms:eu-west-1:123456789012:alias/team-foo
+          # Regional keys of team-bar.
+          - allOf:
+              - match: arn:aws:kms:eu-central-1:123456789012:alias/team-bar
+              - match: arn:aws:kms:eu-west-1:123456789012:alias/team-bar
+      # The AWS KMS key pair of exactly one deployment target environment must
+      # be present.
+      - oneOf:
+          # Regional production keys.
+          - allOf:
+              - match: arn:aws:kms:eu-central-1:123456789012:alias/production-cicd
+              - match: arn:aws:kms:eu-west-1:123456789012:alias/production-cicd
+          # Regional staging keys.
+          - allOf:
+              - match: arn:aws:kms:eu-central-1:123456789012:alias/staging-cicd
+              - match: arn:aws:kms:eu-west-1:123456789012:alias/staging-cicd
+```
+
+### Output
+
+To be as helpful as possible to user, the output should make it clear what the
+problem is. It should include information about:
+
+- The SOPS file that failed the compliance check.
+- Display information about expected, but missing trust anchors ("any of" / "all of").
+- Display information about conflicting trust anchors ("one of").
+- Display information about explicitly denied trust anchors ("not").
+- Display the trust anchors that did not match any rule.
+- Ideally, include context about the line and column in the SOPS file where the
+  check failed.
+- Optionally provide guidance how to make the file compliant with the rules.
```

Comment:
Maybe the rules file should support adding links to documentation? For example like this:

```yaml
rules:
  - allOf:
    - match: age1u79ltfzz5k79ex4mpl3r76p2532xex4mpl3z7vttctudr6gedn6ex4mpl3
      hint: AGE key for offline disaster recovery is required, see https://some-url.com/dr for more information
```

Reply:
Maybe we need both? `description` for the human readable explanation and `url` for the link? That way a UI could decide how to present the information.

Reply:
I'll only add `description` for now, but added a point for "more rule metadata" like URLs or tags/keywords under "Other Considerations" for future reference.